server/models/Classroom.coffee

```diff
@@ -44,13 +45,74 @@ ClassroomSchema.pre('save', (next) ->
     next()
 )
 
-
 ClassroomSchema.methods.isOwner = (userID) ->
   return userID.equals(@get('ownerID'))
 
 ClassroomSchema.methods.isMember = (userID) ->
   return _.any @get('members') or [], (memberID) -> userID.equals(memberID)
 
+ClassroomSchema.methods.generateCoursesData = co.wrap (isAdmin=false) ->
+  # Helper function for generating the latest version of courses
+  query = {}
+  query = {releasePhase: 'released'} unless isAdmin
+  courses = yield Course.find(query)
+  courses = Course.sortCourses courses
+  campaigns = yield Campaign.find({_id: {$in: (course.get('campaignID') for course in courses)}})
+  campaignMap = {}
+  campaignMap[campaign.id] = campaign for campaign in campaigns
+  classLanguage = @get('aceConfig')?.language
+  coursesData = []
+  for course in courses
+    courseData = { _id: course._id, levels: [] }
+    campaign = campaignMap[course.get('campaignID').toString()]
+    levels = _.values(campaign.get('levels'))
+    levels = _.sortBy(levels, 'campaignIndex')
+    for level in levels
+      continue if classLanguage and level.primerLanguage is classLanguage
+      levelData = { original: mongoose.Types.ObjectId(level.original) }
+      _.extend(levelData, _.pick(level, 'type', 'slug', 'name', 'practice', 'practiceThresholdMinutes', 'primerLanguage', 'shareable'))
+      courseData.levels.push(levelData)
+    coursesData.push(courseData)
+  coursesData
+
+ClassroomSchema.methods.generateCourseData = co.wrap (courseId) ->
+  # Helper function for generating the latest version of a course
+  course = yield Course.findById(courseId)
+  campaign = yield Campaign.findById({_id: course.get('campaignID')})
+  classLanguage = @get('aceConfig')?.language
+  courseData = { _id: course._id, levels: [] }
+  levels = _.values(campaign.get('levels'))
+  levels = _.sortBy(levels, 'campaignIndex')
+  for level in levels
+    continue if classLanguage and level.primerLanguage is classLanguage
+    levelData = { original: mongoose.Types.ObjectId(level.original) }
+    _.extend(levelData, _.pick(level, 'type', 'slug', 'name', 'practice', 'practiceThresholdMinutes', 'primerLanguage', 'shareable'))
+    courseData.levels.push(levelData)
+  courseData
+
+ClassroomSchema.methods.setUpdatedCourse = co.wrap (courseId) ->
+  # Update existing or add missing course
+  latestCourse = yield @generateCourseData(courseId)
+  updatedCourses = _.clone(@get('courses') or [])
+  existingIndex = _.findIndex(updatedCourses, (c) -> c._id.equals(courseId))
+  if existingIndex >= 0
+    updatedCourses.splice(existingIndex, 1, latestCourse)
+  else
+    # TODO: does this need to be inserted in order?
+    updatedCourses.push(latestCourse)
+  @set('courses', updatedCourses)
+
+ClassroomSchema.methods.setUpdatedCourses = co.wrap (isAdmin=false, addNewCoursesOnly=true) ->
```

Comment:
Second argument should be the inverse, for readability: eg `updateOldCourses`. Naming things with the fewest negatives is the least confusing IMO.